### ISSUE TITLE
Adding @react-native-firebase breaks incoming calls on Android

### ISSUE THREAD

**Issue by clundberg-nextiva** [twilio/twilio-voice-react-native#370]
# Issue

## Pre-submission Checklist
- [x] I have verified that the issue occurs with the latest release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-voice-react-native/blob/main/CHANGELOG.md).
- [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-voice-react-native/blob/main/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
- [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

## Description 
After adding `@react-native-firebase/app` and `@react-native-firebase/messaging`, it seems that the CallInvite event is never emitted.

## Reproduction Steps
1. Add `@react-native-firebase/app` and `@react-native-firebase/messaging`
2. Receive an incoming call

## Expected Behavior
When receiving an incoming call, the `CallInvite` callback should be invoked and a notification should be shown.

```
voice.on(Voice.Event.CallInvite, (callInvite: CallInvite) => {
  console.log('call invite', callInvite);
});
```

## Actual Behavior
Nothing. The callback is never invoked and the incoming call notification never appears.

## Reproduction Frequency
100%

## Software and Device Information
**Please complete the following information.**
 - Device: Pixel 3a
 - OS: Android API Level 34
 - React version: 18.2.0
 - React Native version: 0.73.6
 - Node version: 18.20.2
 - npm or yarn version: npm v10.5.0

## Additional Context
I can also confirm that `firebase.messaging().getToken()` and `new Voice().getDeviceToken()` are returning the same token, and if I add the firebase `onMessage` handler, I can also see that I'm receiving the Twilio message, but it doesn't seem to be processed by the SDK.

```
useEffect(() => {
  const unsubscribe = firebase.messaging().onMessage((message) => {
    console.log('remote message', message);
  });

  return unsubscribe;
}, []);
```

**Comment by afalls-twilio** [comment]
@clundberg01  What happens when you delete the app and reinstall. If both our firebase module and yours use the same token, they should both work in tandem.

**Comment by clundberg01** [comment]
@afalls-twilio The same issue continues to happen, even after wiping the emulator.

**Comment by afalls-twilio** [comment]
@clundberg01 what version of our sdk are you using?

**Comment by shrutigarg-reactnative** [comment]
Same issue in release mode , for debug it works fine

**Comment by ilkinabdullayev** [comment]
i have same issue :(

**Comment by ariloc** [comment]
Hi, I've been experiencing the same issue, also in SDK version 1.0.0. After some searching, I've realized it's related to this issue in the `react-native-firebase` repo: https://github.com/invertase/react-native-firebase/issues/1129

Diving in the native modules, both the `@react-native-firebase/messaging` and the `@twilio/voice-react-native-sdk` seem to declare a service with the same `intent-filter`:

https://github.com/twilio/twilio-voice-react-native/blob/5b4c0e74676e7c62722590e61379e6776f8ae114/android/src/main/AndroidManifest.xml#L13

https://github.com/invertase/react-native-firebase/blob/3fa3f110b357ef0dbe2cc0fc12e982edc913b588/packages/messaging/android/src/main/AndroidManifest.xml#L15

As far as I understand, when compiling, push notifications can only be delivered to one of these services, causing issues such as incoming voip calls to not be received by the library.

I've followed some of the recommendations suggested in the previously mentioned issue, and could partially fix it by creating my own service:

```kt
class CustomFirebaseMessagingService : FirebaseMessagingService() {
    private val rnFirebaseService: ReactNativeFirebaseMessagingService =
        ReactNativeFirebaseMessagingService()
    private val twilioVoipService: VoiceFirebaseMessagingService =
        VoiceFirebaseMessagingService()

    override fun onMessageReceived(message: RemoteMessage) {
        // TODO: Add same wakelock as service
        if (message.getData().isEmpty() ||
            !Voice.handleMessage(
                this,
                message.getData(),
                twilioVoipService,
                CallMessageListenerProxy()
            )) {
            // Trigger RNFirebase message if not a valid Twilio Voip Push Notification
            // (currently noop though, because it is currently handled by a different, deprecated
            // intent filter, com.google.android.c2dm.intent.RECEIVE, see React Native
            // Firebase AndroidManifest.xml)
            rnFirebaseService.onMessageReceived(message)
        }
    }

    override fun onNewToken(token: String) {
        twilioVoipService.onNewToken(token)
        rnFirebaseService.onNewToken(token)
    }

    override fun onMessageSent(msgId: String) {
        twilioVoipService.onMessageSent(msgId)
        rnFirebaseService.onMessageSent(msgId)
    }

    override fun onDeletedMessages() {
        twilioVoipService.onDeletedMessages()
        rnFirebaseService.onDeletedMessages()
    }

    override fun onSendError(msgId: String, exception: Exception) {
        twilioVoipService.onSendError(msgId, exception)
        rnFirebaseService.onSendError(msgId, exception)
    }
}
```

This service has to be declared in my app as:

```xml
<service android:name=".CustomFirebaseMessagingService"
  android:exported="false">
    <intent-filter>
        <action android:name="com.google.firebase.MESSAGING_EVENT" />
    </intent-filter>
</service>
```

And the following dependencies have to be added to the app's `build.gradle` file (using `@react-native-firebase/messaging` version 20.1.0):

```gradle
// CustomFirebaseMessagingService
implementation "com.twilio:voice-android:6.4.1"
implementation "com.google.firebase:firebase-messaging:23.4.0"
```

Again, this is a partial fix. For instance, I have not included the code responsible of waking the device ([here](https://github.com/twilio/twilio-voice-react-native/blob/5b4c0e74676e7c62722590e61379e6776f8ae114/android/src/main/java/com/twiliovoicereactnative/VoiceFirebaseMessagingService.java#L42)), although it seems it would be pretty straightforward to include it, using the `Context` from the service I created.

Also, as indicated by my comment in the code, and as explained [here](https://github.com/invertase/react-native-firebase/discussions/6090), the `@react-native-firebase/messaging` library currently uses a separate intent filter for receiving messages, and actually seems it isn't being used in the main service (see [here](https://github.com/invertase/react-native-firebase/blob/3fa3f110b357ef0dbe2cc0fc12e982edc913b588/packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingService.java#L34)), so there should be no need to call that event from my service. I opted to do it anyways in case the `@react-native-firebase/messaging` library is updated in the future.
That's also the same reason I called the event for functions `onMessageSent`, `onDeletedMessages`, `onSendError` in the voice messaging service, given that they seem to be unused currently ([here](https://github.com/twilio/twilio-voice-react-native/blob/5b4c0e74676e7c62722590e61379e6776f8ae114/android/src/main/java/com/twiliovoicereactnative/VoiceFirebaseMessagingService.java#L37)).

Note that I haven't tested all the features fully, nor have I looked at all the code to check if my changes affected it. Nevertheless, I wanted to note that the major issue I encountered is related to having a reference to the `Context`, as explained in [this article](https://medium.com/@chornenkyy.v/do-you-want-to-use-two-or-more-firebasemessagingservices-at-once-67236d1bd9e0), linked by [this answer](https://github.com/invertase/react-native-firebase/issues/1129#issuecomment-479058235) in the issue already mentioned. 

I just wanted to post my solution in case someone else finds it useful, as well as to help maintainers seek for an official solution to this issue. The need of managing more push notifications in an app,than the ones used by the voice functionality, doesn't seem unreasonable, and if someone knows of a better solution I overlooked, I'd really appreciate it.

Additionally, I looked if there were any recommendations for managing other push notifications in the [Android quick start repo](https://github.com/twilio/voice-quickstart-android), considering the first major release of the React Native SDK is really recent, and that the Android SDK development may be more mature. However, I didn't find any, but again, please let me know if I missed anything.

**Comment by mhuynh5757** [comment]
Hi @ariloc thanks for putting in the research and taking the time to make this extensive write-up, I really appreciate it! Could you let me know what versions you're using, React Native, Android, etc.? I did some testing with a fresh React Native app myself with both `@react-native-firebase/messaging` and our library, and I'm not able to reproduce this issue. I'm wondering if it's version dependent.

**Comment by hieulamblackbird** [comment]
I'm currently having the same issue. I've tested it with my Expo project running on an Android 13 phone. Here are the version numbers:

```json
"expo": "~51.0.21",
"expo-router": "~3.5.18",
"react-native": "0.74.3",
"@react-native-firebase/app": "~20.2.1",
"@react-native-firebase/messaging": "~20.2.1",
```

**Comment by ariloc** [comment]
> Hi @ariloc thanks for putting in the research and taking the time to make this extensive write-up, I really appreciate it! Could you let me know what versions you're using, React Native, Android, etc.? I did some testing with a fresh React Native app myself with both `@react-native-firebase/messaging` and our library, and I'm not able to reproduce this issue. I'm wondering if it's version dependent.

Thank you for your quick response @mhuynh5757 . The project in which I was integrating this library is running `react-native` version 0.74.0, and I was testing the app in an Android emulator running API 33 (Android 13). As for the versions of the libraries, as I mentioned in my previous comment, are 1.0.0 for `@twilio/voice-react-native-sdk` and 20.1.0 for `@react-native-firebase/messaging`. I'm also running `@react-native-firebase/app` and `@react-native-firebase/auth` with the same version in that project. Note that I'm not using Expo.

If the issue is the one I mentioned, but you couldn't reproduce it, it's possible that FCM messages are being received by the `VoiceFirebaseMessagingService` instead of the `ReactNativeFirebaseMessagingService` through the `com.google.firebase.MESSAGING_EVENT` intent filter, but messages could still be received by the `@react-native-firebase/messaging` library because they are handled separately with another intent filter. However, that's not the case for other events such as `onDeletedMessages`, which could be tested: https://rnfirebase.io/reference/messaging#onDeletedMessages.
I guess it depends on which service Android chooses to deliver the intents to, however I'm not sure how to determine that.

I also tried to create a new project from scratch to test if the issue is still present, in case any other libraries I was using in the project were causing the issue. After creating a new blank non-Expo project, I confirmed that the issue persists, and I'm only able to receive calls if I remove the `@react-native-firebase/app` and `@react-native-firebase/messaging` libraries. These are the only dependencies I used in the new project:

```json
"@react-native-firebase/app": "^20.3.0",
"@react-native-firebase/messaging": "^20.3.0",
"@twilio/voice-react-native-sdk": "^1.0.0",
"react": "18.2.0",
"react-native": "0.74.3"
```

If you need any more information about the issue, please let me know.

**Comment by mhuynh5757** [comment]
Thanks @ariloc for providing more info. I've forwarded this to my team and we'll try to prioritize looking into this.

**Comment by afalls-twilio** [comment]
Hi all, (@hieulamblackbird @ariloc @ilkinabdullayev etc..) this thread has turned into a dogpile of potentially different issues...


So let me try to break this down...

**The problem.**

The reason you are all having issues is because the SDK has a FirebaseMessaging Service embedded inside of it so that it can receive incoming calls. When using `@react-native-firebase/messaging` you are creating a potential conflict where now two services are listening to firebase messaging events. 

From our internal testing, if both services reacting to firebase messages are using the same firebase token, then both services will receive the same message. The firebase token is created and registered by the Firebase SDK the first time the application is ran. The problem occurs when multiple services are using different tokens. This can happen several ways, the most common is that `@react-native-firebase/messaging` package is added to your project after some development has already occurred where a version of your application was ran without containing it. This is because the embedded service in our Voice SDK has created a token and registered it and then when the `@react-native-firebase/messaging` package was added later and then the subsequent updated app was ran, the ``@react-native-firebase/messaging` created and registered a new firebase token.


**The Solution**

There are several solutions, I'll try to list them here...

1) Completely delete your application from the devices you are testing on, including your emulator.  Delete the application by long-pressing the application icon, then selecting 'uninstall application'. If you use `adb install -r` or press 'run' from Android Studio, this will not delete all the data and will not force the generation of a new FCM token. Then install your application and run it.

2) This solution requires some java knowledge but because the react native voice SDK is open source, you can take the source and modify the code. The [VoiceFirebaseMessagingService](https://github.com/twilio/twilio-voice-react-native/blob/main/android/src/main/java/com/twiliovoicereactnative/VoiceFirebaseMessagingService.java) is where the embedded firebase service lives in the the React Native Android Voice SDK. You can modify this to also trap additional firebase push messages unique to your application.

**Debugging Tips**

Check that both tokens from `@react-native-firebase/messaging` and from [VoiceFirebaseMessagingService](https://github.com/twilio/twilio-voice-react-native/blob/main/android/src/main/java/com/twiliovoicereactnative/VoiceFirebaseMessagingService.java)  are the same. If they are not, that is the root of your issue.

**Resources**

More about Firebase Message Tokens...
https://firebase.blog/posts/2023/04/managing-cloud-messaging-tokens/

**Comment by ariloc** [comment]
Hi @afalls-twilio. thank you for your detailed response. I saw some related issues where you were asking if the tokens matched, so now I understand the reason behind those questions. 

Is this behavior documented by Google? I couldn't find any mention of it specifically, but I understand that two devices with the same token should receive the same message.
I tried your first solution on the new project I previously created, and made sure both tokens were the same (using `getToken()` from `@react-native-firebase/messaging` and `getDeviceToken()` from the Voip SDK). However, the reported behavior didn't change: messages were only received by the  `@react-native-firebase/messaging` library, and no calls were received. Again, just for good measure, I tried removing the `@react-native-firebase/messaging` and `@react-native-firebase/app` libraries afterwards, and receiving calls was back working again.

I also checked if messages were received at all by the `VoiceFirebaseMessagingService` when both dependencies were installed and the tokens matched, by checking the logs in Android Studio, but it didn't seem like they were. Note that these logs were actually present when I removed the `@react-native-firebase` libraries.

Moreover, I noticed messages such as "Successfully registered FCM" from `TwilioVoiceReactNativeModule` were logged with both dependencies installed and tokens matching, but then only messages from `RNFirebaseMsgReceiver` were logged when I placed a call (and therefore, a FCM notification was received). 
If messages were only received by only one of the services as I initially thought, then the `onNewToken()` event should only be called in one of the services (`ReactNativeFirebaseMessagingService` and `VoiceFirebaseMessagingService`). I added a log line of my own in the `ReactNativeFirebaseMessagingService` for the aforementioned event (as I couldn't find if it already had one that was triggered directly by the event), and the following messages were logged around one second of difference apart from the other:

```
RNFIREBASE  com.twiliovoicetest  D  ON NEW TOKEN [my token]
[...]
TwilioVoic...tiveModule  com.twiliovoicetest  D  Registering with FCM with token [my token]
```
(both tokens are matching)

The first one is the line I placed myself, indicating the event was triggered in `ReactNativeFirebaseMessagingService`. The second one is from `TwilioVoiceReactNativeModule`, but doesn't correspond to the event being triggered in the `VoiceFirebaseMessagingService`, because as you can see in the [code](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/android/src/main/java/com/twiliovoicereactnative/VoiceFirebaseMessagingService.java#L28), if the `onNewToken()` event was triggered, the message logged would start like `Refreshed FCM token: ...`. This message actually corresponds with [this line in the code](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java#L445), which I think is triggered because I'm registering to receive calls in the React Native code with `voice.register(...);`. Afterwards, the previously mentioned "Successfully registered FCM" message is logged, because the registration process is attached to a listener [here](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java#L446) and the message is logged [here](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java#L123).
Note that the token used in the `TwilioVoiceReactNativeModule.java` file is retrieved from the current [`FirebaseMessaging`](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessaging) instance [here](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java#L233). 
When I looked for the message that should be logged by `onNewEvent()` triggering in the `VoiceFirebaseMessagingService` I couldn't find it, thus I think showing that only one of the services, the `ReactNativeFirebaseMessagingService` in this case, is receiving all FCM messages.

So it seems that both services can use the same token, but the conflict between the services still be present, at least according to my testing.

About your second solution, using a third service, although not ideal, is a way of handling all messages needed by my app manually. I opted for this solution because usually, and at least in my case, files downloaded by dependencies aren't committed to the repository, so I would have to make an exception in my `.gitignore` file to account for this case, and performing any updates to the dependencies may overwrite my modifications.

I hope this information helps with understanding the issue. I don't know if there is a proper way to fix it, but using the third service seems to be, at the moment, the only solution I found to be working. Therefore, I just wanted to report my findings so maybe an official and tested solution could be supported and documented.
As a bit of context, the reason I'm using the `@react-native-firebase/messaging` library in the first place is because I needed to handle push notifications from Twilio Conversations, which I'm currently using in a project through the JS SDK, which as far as I could gather, [it is the recommended approach for using the service in React Native](https://stackoverflow.com/questions/66250024/is-there-a-library-to-use-twilio-conversations-in-react-native).

**Comment by afalls-twilio** [comment]
@ariloc what does the merged android manifest look like? Do both services show up? (1 from `@react-native-firebase/messaging` and `VoiceFirebaseMessagingService`)

**Comment by ariloc** [comment]
@afalls-twilio Here's the merged Android manifest for the blank project I created previously (found in the `merged_manifest` folder, inside the build folder as you suggested):

[AndroidManifest.xml.gz](https://github.com/user-attachments/files/16434994/AndroidManifest.xml.gz)

I zipped it with `gzip` because GitHub didn't allow me to upload it a file with a `.xml` extension, and it's quite long to be pasting it here.

Seems like both services show up, and have the `com.google.firebase.MESSAGING_EVENT` intent filter. There's actually three services with it, but the third one seems to be a native Firebase library from Google, with an `android:priority` attribute set to -500.
I think it may be relevant to note that the service from `@react-native-firebase/messaging` is listed above the `VoiceFirebaseMessagingService`.

When I first looked into this issue, I wondered if two services with the same intent-filter could receive the same intents. Most answers I found indicated that they couldn't, but I didn't find anything clear about the priorities with which services received intents if two or more had a matching intent-filter. Apart from a [StackOverflow answer that said it was random](https://stackoverflow.com/questions/8381324/two-service-with-the-same-intent-filter), [this](https://developer.android.com/guide/topics/manifest/intent-filter-element) is the best I could find from the official documentation.
The attribute `android:order` should be useful to "desambiguate multiple matching filters in a single app", just what we're looking for, though it was introduced in Android API 28. It also doesn't clarify how intents would be delivered to two services with the same value for this attribute (the default, in this case).
On the other hand, `android:priority` seems to have a similar purpose, and I also found it mentioned in some of the answers I previously got. Nevertheless, it differs from the other attribute in that "priority applies across apps, while order disambiguates (...) in a single app". No clarification of behavior when two services have same priority, either.

So, as much as the documentation concerns (and what I could find), this behavior seems to be undefined, and it is definitely not random (or at least not in the way I was thinking of, otherwise Firebase messages should _sometimes_ be received by the `VoiceFirebaseMessagingService`).
My guess is that the order services are merged into the manifest may be relevant to the issue, therefore explaining that the `@react-native-firebase/messaging` being listed first in the merged manifest is the reason it's receiving all Firebase messaging intents. Even if this isn't the case, I shouldn't be depending on undefined behavior for my project to work.
Still, under that assumption, I also took another assumption, in that project specific services may have more priority that the ones defined in external libraries. At least, they are prioritized like this when the manifests are merged, [according to the docs](https://developer.android.com/build/manage-manifests#merge_priorities). 
I actually just checked the merged manifest in the project where I used a third service, and it's listed above both services (the one from `@react-native-firebase/messaging` and the `VoiceFirebaseMessagingService`).
Because of this, and partly because of an answer I found in another discussion (which I mentioned in my initial comment and suggested creating a third service as a solution), is why I decided to solve it this way, and the reason it makes sense to me.

Thinking about it now, I tried setting the `android:order` attribute in the intent-filter of the `VoiceFirebaseMessagingService` to a higher value than default (I chose 10), to check if the reported behavior changed. Sure enough, calls were being received correctly now. However, checking the merged manifest, the order of the declared services didn't change, as the service from `@react-native-firebase/messaging` was still being listed above the `VoiceFirebaseMessagingService`. The only difference in this manifest was my change to the `android:order` attribute.
Note that if I removed the attribute and recompiled the app, the issue was back, without even clearing the cache in Android Studio, or previously uninstalling the app (and the tokens were matching each time).

Similarly, I tried setting the `android:priority` attribute, but this time with a negative value on the intent-filter of the service from `@react-native-firebase/messaging`. I did it this way since, from what I understand of the documentation, the attribute is ignored if a non-privileged app uses a value higher than 0, while also the default value *is* 0. Therefore, the only way I could prioritize the `VoiceFirebaseMessagingService` is instead by lowering the priority of the other service (I chose -10). Again, this fixed the issue and receiving calls was working again. The order of the declared services in the merged manifest remained the same (the attribute being the only one difference I could notice), and the issue also persisted after removing the attribute.

Finally, I checked if the token was the same in all three services in my other project, as you requested. All three are matching. I tested this by adding a log line in the `onNewToken()` event of my custom service, and using the same functions from each library I used in my previous comment.

**Comment by afalls-twilio** [comment]
@ariloc Thanks for providing your manifest... Interesting information about service order...  internally we are putting a demo of it working (on our end) together that we can share so that we can go back and forth with you until we both get the same behavior.

**Comment by mhuynh5757** [comment]
@ariloc thanks for the great community conversation. I hope that we can get to the bottom of this as soon as possible.

I attempted to reproduce the issue using the embedded test app that's in this repo, under `test/app`. I added the `@react-native-firebase/app` and `@react-native-firebase/messaging` packages to the test app and unfortunately I don't see the issue such that call invites are not received by the test app. In fact, I see that the call invite Firebase message is received by the RNFirebase messaging package, as well as the Twilio SDK and I can interact with the call invite as usual, accepting or rejecting it.

The work is in a test branch called `test/rnfirebase-messaging` link here for convenience (https://github.com/twilio/twilio-voice-react-native/tree/test/rnfirebase-messaging). Do you think you could try building and running this app with your dev environment to see if you're encountering what I am, or still encountering your issue?

To run the test app, you need to provide a file `src/e2e-tests-token.ts` or otherwise change the `voice.connect`/`callInvite.accept` code to provide a token for your use case. The `src/e2e-tests-token.ts` file can look like this:

```
export const token = 'foobar';
```

Please let me know if this sounds reasonable to you. Thanks again for helping us to resolve this together.

**Comment by ariloc** [comment]
@mhuynh5757 Sure, no problem.
I figured I had to ran `yarn app` first in the repo root folder to get the app running, as it seems to be generating some constant files that were required by the Twilio Voice React Native library. 
Indeed, after registering inside the app you provided, I can receive calls just fine.

I checked the generated merged manifest and the `VoiceFirebaseMessagingService` is listed above the service from `@react-native-firebase/messaging`, while in my merged manifest from my own project it was the other way.
Reading through the [React Native documentation](https://reactnative.dev/docs/native-modules-android?android-language=kotlin), I noticed that packages are added to the project in the `MainReactNativeHost` file at the root. Though, due to [autolinking](https://github.com/react-native-community/cli/blob/main/docs/autolinking.md), a `PackageList.java` file is generated with the package list for packages that support it. I also noticed that the order of the libraries inside the `getPackages()` function in this file, is different between the projects. While the generated file in the project you provided looks like this:

```java
public ArrayList<ReactPackage> getPackages() {
    return new ArrayList<>(Arrays.<ReactPackage>asList(
      new MainReactPackage(mConfig),
      new TwilioVoiceReactNativePackage(),
      new ReactNativeFirebaseAppPackage(),
      new ReactNativeFirebaseMessagingPackage()
    ));
  }
```

In my project it looks like this:

```java
public ArrayList<ReactPackage> getPackages() {
    return new ArrayList<>(Arrays.<ReactPackage>asList(
      new MainReactPackage(mConfig),
      new ReactNativeFirebaseAppPackage(),
      new ReactNativeFirebaseMessagingPackage(),
      new TwilioVoiceReactNativePackage()
    ));
  }
```

Looks like that, in your project, the generated file lists the `TwilioVoiceReactNativePackage` before the `ReactNativeFirebaseAppPackage` and the `ReactNativeFirebaseMessagingPackage`, while it's the other way around in mine.
This may affect the order of the services in the merged manifest (which, considering my assumption from my previous comment is correct, it may affect the priority with which services receive events).

[The method that creates the mentioned file is called `applyNativeModulesAppBuildGradle`](https://github.com/react-native-community/cli/blob/main/docs/autolinking.md#platform-android), which is defined [here](https://github.com/react-native-community/cli/blob/6a61d5dcc99c6430b74c1095e05ffb315909b078/packages/cli-platform-android/native_modules.gradle#L551). I'm not personally familiar with Gradle build scripts, though at first glance, seems that the `PackagesList.java` file is generated with [this function](https://github.com/react-native-community/cli/blob/6a61d5dcc99c6430b74c1095e05ffb315909b078/packages/cli-platform-android/native_modules.gradle#L223) and may just join packages in order, which are taken from the variable `this.reactNativeModules`, [received by the class that contains the function in its constructor](https://github.com/react-native-community/cli/blob/6a61d5dcc99c6430b74c1095e05ffb315909b078/packages/cli-platform-android/native_modules.gradle#L158). However, I couldn't get any strong conclusions from this.

Another thing I also just spotted, maybe by lack of knowledge, is that the Twilio Voice SDK package dependency isn't listed within the `packages.json` file of the project. This is clearly done to test the module from the local version, so it's not a surprise given you use this app for testing it in your repo. If I'm not mistaken, the dependency is actually declared in the `react-native.config.js` file [here](https://github.com/twilio/twilio-voice-react-native/blob/3ed747373e2e5fe2cf59288d8cf3c5386d508785/test/app/react-native.config.js), so I wondered if the packages listed here may be prioritized over the dependencies from `yarn`.
Therefore, I tried linking the package using `yarn` instead, as suggested in the [React Native documentation for linking local native modules](https://reactnative.dev/docs/local-library-setup). Specifically, I commented the contents of `module.exports` in `react-native.config.js` like this:

```js
const path = require('path');
const pak = require('../../package.json');

module.exports = {
  /*
  dependencies: {
    [pak.name]: {
      root: path.join(__dirname, '../..'),
    },
  },
  */
};
```

And then added the dependency to the `package.json` of the app adding the line `"@twilio/voice-react-native-sdk": "link:../../"` as follows:

```json
"dependencies": {
    "@react-native-firebase/app": "^20.3.0",
    "@react-native-firebase/messaging": "^20.3.0",
    "@twilio/voice-react-native-sdk": "link:../../",
    "react": "18.2.0",
    "react-native": "0.72.9"
  },
```
(adding no spaces between `link:` and the path `../../` is crucial, or at least I experienced issues if I used `link: ../../` instead)

This creates a symlink in `./node_modules/@twilio/voice-react-native-sdk` to the root folder of the repo, so if you experience issues with the linking not being done properly, using the absolute path of the folder in your file system may help.
  
After clearing the `yarn` cache and Android Studio's cache as well (just in case), I could compile the app successfully again. Now, the newly generated `PackageList.json` file looks identical to the one in my project (the one I showed before), and the services in the merged manifest are listed more similarly to how they are in my project: the service from `@react-native-firebase/messaging` is declared above the `VoiceFirebaseMessagingService`.
Testing receiving calls now, it's no longer working. However, messages are being received in the `@react-native-firebase/messaging` service, as they are logged in the console when I try to call the device.

Please check if you can make the changes I made to the repo and replicate the described behavior.

**Comment by ariloc** [comment]
Hi @mhuynh5757, do you have any updates on the issue? I saw you created PR #394 which could help with it, allowing me to handle the push notifications received by the service from `@react-native-firebase/messaging` and passing them to the Twilio Voice Library within React Native, while only having a single service with the `com.google.firebase.MESSAGING_EVENT` intent filter. Would you like me to test it?

**Comment by mhuynh5757** [comment]
Hi @ariloc if you want to test the branch, that would be great. I had some other priorities I needed to address but I will be getting started soon on another demo app to attempt to reproduce the error. I suspect that you're on to something regarding the manifest ordering, and I suspect it indeed has something to do with the autolinking.

**Comment by ariloc** [comment]
No problem @mhuynh5757, when you have any further updates please let me know!

About testing the branch, I had to add a few lines to some interfaces and classes in the library so I could use the new `handleEvent` function in the API.

At `src/type/NativeModule.ts`:
```diff
...
  voice_showNativeAvRoutePicker(): Promise<void>;
  voice_unregister(accessToken: string): Promise<void>;
+ voice_handleEvent(
+   messageData: Record<string,any>
+ ): Promise<boolean>;
}
```

At `src/Voice.tsx`:
```diff
...
  unregister(token: string): Promise<void> {
    return NativeModule.voice_unregister(token);
  }

+ handleEvent(messageData: Record<string,any>): Promise<boolean> {
+   return NativeModule.voice_handleEvent(messageData)
+ }
...
```

Then, as described in the PR, I added the property value to `config.properties` in the assets directory. I also removed the third service from my project.

Now, for handling events, I can receive messages using `@react-native-firebase/messaging` handlers like `onForeground` or `onBackground`, and pass them to the Twilio Voip SDK with the new function. I can also check which messages should be passed to the Twilio Voip SDK and which shouldn't, by checking the format of the received message. In my case, checking if the received data has a field called `twi_message_type` with value `twilio.voice.call` is enough for this purpose.
Finally, it seems like I can just pass the received `message.data` to the new `handleEvent` function and works like intended! (where `message` is `FirebaseMessagingTypes.RemoteMessage`):

```tsx
return await voice.handleEvent(message.data)
```

Although I tested that both receiving calls and other kinds of push notifications were received and handled correctly, I noticed that the callbacks for events like receiving a new CallInvite (`Voice.Event.CallInvite`, which I was previously using to redirect the user to a call screen when they answered a call), were no longer being called.

I looked at the code, digging a bit into how Native Modules work in React Native and how the documentation recommended integration with Android ([here](https://reactnative.dev/docs/native-modules-android)), but I couldn't find anything that seemed off between the previous version of my app (the one I was using a third service) and the new one.
I checked if using `reactContext` instead of the service context as a parameter for `Voice.handleMessage()` made any differences, but couldn't find anything of relevance.
I also tried adding the listeners for the aforementioned events in other parts of the app within the React Native code, in case the one I set up when starting the app somehow got invalidated, but I couldn't get it to work either.

Events seem to be sent correctly at `JSEventEmitter.java`, which writes lines like "`sendEvent scopeVoice params {"callInvite": {[REDACTED]},"type":"voiceEventTypeValueIncomingCallInvite"}}`". There are ocasionally some warnings such as "`attempt to sendEvent without context {com.facebook.react.bridge.ReactApplicationContext@48f1fd} or Catalyst instance not active`", but I noticed that these are also written to the logs in the previous version of the app with the third service, where callbacks are working correctly.

**Comment by mhuynh5757** [comment]
Hi @culov I'm really sorry that this is blocking you, the team is aware and we're trying to prioritize a solution that works for all use cases. Thanks for your patience, I'll let the team know.

**Comment by mhuynh5757** [comment]
@culov if you are willing to test that PR, the team would appreciate any and all feedback. The reason we can't merge that PR as-is is because it would be a breaking change in default behavior.

**Comment by LukakHR** [comment]
@mhuynh5757 will this be merged soon? Our app's incoming calls aren't working because of this

**Comment by mhuynh5757** [comment]
Hi all, sorry for the delays regarding this fix. You have my sincerest apologies. We're close to merging the fix to `main`, the branch that has this fix is `task/disable-firebase-service`. Once that's merged to `main` we're planning on cutting a release that includes it this week. If you are feeling adventurous and want to try out the changes beforehand, feedback would be greatly welcomed. Thanks all for your patience!

**Comment by aaronleo28** [comment]
@mhuynh5757 Great news on the fix, thanks for all your efforts with this work

do you have any further updates on when this will be released?

**Comment by mhuynh5757** [comment]
I actually do have some updates 🙂 it's out now! 🎉 

`1.2.1` was released this morning and includes the fix for this issue. Please upgrade at your earliest convenience. I want to thank everyone who participated in the discovery and research for this issue, and thank you all for your patience while we were busy working on it.

**Comment by mhuynh5757** [comment]
I'll close this issue for now as resolved, but if anyone has any issues with the fix using `1.2.1` please reach out for further assistance.
